File: acgs2-core/sdk/typescript/src/services/governance.ts
/**
 * ACGS-2 Governance Service
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import { ACGS2Client } from '../client';
import {
  CONSTITUTIONAL_HASH,
  GovernanceDecision,
  GovernanceDecisionSchema,
  ApprovalRequest,
  ApprovalRequestSchema,
  ApprovalStatus,
  CreateApprovalRequest,
  SubmitApprovalDecision,
  PaginationParams,
  PaginatedResponse,
} from '../types';
import { ValidationError, nowISO } from '../utils';

// =============================================================================
// Governance Types
// =============================================================================

export interface GovernancePolicy {
  id: string;
  name: string;
  description: string;
  type: 'constitutional' | 'operational' | 'ethical';
  priority: number;
  enabled: boolean;
  rules: GovernanceRule[];
  constitutionalHash: string;
}

export interface GovernanceRule {
  id: string;
  name: string;
  condition: string;
  action: 'allow' | 'deny' | 'require_approval';
  escalationLevel?: number;
  metadata?: Record<string, unknown>;
}

export interface EscalationPath {
  id: string;
  name: string;
  levels: Array<{
    level: number;
    approverRoles: string[];
    timeoutMinutes: number;
    autoEscalate: boolean;
  }>;
  constitutionalHash: string;
}

export interface GovernanceMetrics {
  totalDecisions: number;
  approvedCount: number;
  deniedCount: number;
  escalatedCount: number;
  averageDecisionTimeMs: number;
  complianceRate: number;
  byPolicy: Record<string, { decisions: number; complianceRate: number }>;
  constitutionalHash: string;
}

// =============================================================================
// Governance Service
// =============================================================================

export class GovernanceService {
  private readonly basePath = '/api/v1/governance';

  constructor(private readonly client: ACGS2Client) {}

  // ===========================================================================
  // Approval Requests
  // ===========================================================================

  /**
   * Creates an approval request
   */
  async createApprovalRequest(request: CreateApprovalRequest): Promise<ApprovalRequest> {
    const response = await this.client.post<ApprovalRequest>(`${this.basePath}/approvals`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to create approval request');
    }

    return this.validateApprovalRequest(response.data);
  }

  /**
   * Gets an approval request by ID
   */
  async getApprovalRequest(requestId: string): Promise<ApprovalRequest> {
    const response = await this.client.get<ApprovalRequest>(
      `${this.basePath}/approvals/${requestId}`
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Approval request not found: ${requestId}`);
    }

    return this.validateApprovalRequest(response.data);
  }

  /**
   * Lists approval requests
   */
  async listApprovalRequests(params?: PaginationParams & {
    status?: ApprovalStatus;
    requesterId?: string;
    pendingFor?: string; // Approver ID
  }): Promise<PaginatedResponse<ApprovalRequest>> {
    const response = await this.client.get<PaginatedResponse<ApprovalRequest>>(
      `${this.basePath}/approvals`,
      params
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list approval requests');
    }

    response.data.data = response.data.data.map((r) => this.validateApprovalRequest(r));
    return response.data;
  }

  /**
   * Submits an approval decision
   */
  async submitDecision(
    requestId: string,
    decision: SubmitApprovalDecision
  ): Promise<ApprovalRequest> {
    const response = await this.client.post<ApprovalRequest>(
      `${this.basePath}/approvals/${requestId}/decisions`,
      {
        ...decision,
        timestamp: nowISO(),
        constitutionalHash: CONSTITUTIONAL_HASH,
      }
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to submit decision for request: ${requestId}`);
    }

    return this.validateApprovalRequest(response.data);
  }

  /**
   * Escalates an approval request
   */
  async escalate(requestId: string, reason: string): Promise<ApprovalRequest> {
    const response = await this.client.post<ApprovalRequest>(
      `${this.basePath}/approvals/${requestId}/escalate`,
      { reason }
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to escalate request: ${requestId}`);
    }

    return this.validateApprovalRequest(response.data);
  }

  /**
   * Cancels an approval request
   */
  async cancelApprovalRequest(requestId: string, reason?: string): Promise<void> {
    await this.client.post(`${this.basePath}/approvals/${requestId}/cancel`, { reason });
  }

  // ===========================================================================
  // Governance Decisions
  // ===========================================================================

  /**
   * Gets a governance decision by ID
   */
  async getDecision(decisionId: string): Promise<GovernanceDecision> {
    const response = await this.client.get<GovernanceDecision>(
      `${this.basePath}/decisions/${decisionId}`
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Governance decision not found: ${decisionId}`);
    }

    return this.validateDecision(response.data);
  }

  /**
   * Lists governance decisions
   */
  async listDecisions(params?: PaginationParams & {
    decision?: 'approve' | 'deny' | 'escalate';
    requestId?: string;
    reviewerId?: string;
  }): Promise<PaginatedResponse<GovernanceDecision>> {
    const response = await this.client.get<PaginatedResponse<GovernanceDecision>>(
      `${this.basePath}/decisions`,
      params
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list governance decisions');
    }

    response.data.data = response.data.data.map((d) => this.validateDecision(d));
    return response.data;
  }

  /**
   * Verifies a governance decision's blockchain anchor
   */
  async verifyDecisionAnchor(decisionId: string): Promise<{
    verified: boolean;
    anchorDetails?: {
      transactionHash: string;
      blockNumber: number;
      timestamp: string;
    };
    constitutionalHash: string;
  }> {
    const response = await this.client.get<{
      verified: boolean;
      anchorDetails?: {
        transactionHash: string;
        blockNumber: number;
        timestamp: string;
      };
      constitutionalHash: string;
    }>(`${this.basePath}/decisions/${decisionId}/verify`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to verify decision anchor: ${decisionId}`);
    }

    return response.data;
  }

  // ===========================================================================
  // Governance Policies
  // ===========================================================================

  /**
   * Lists governance policies
   */
  async listPolicies(params?: PaginationParams & {
    type?: 'constitutional' | 'operational' | 'ethical';
    enabled?: boolean;
  }): Promise<PaginatedResponse<GovernancePolicy>> {
    const response = await this.client.get<PaginatedResponse<GovernancePolicy>>(
      `${this.basePath}/policies`,
      params
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list governance policies');
    }

    return response.data;
  }

  /**
   * Gets a governance policy by ID
   */
  async getPolicy(policyId: string): Promise<GovernancePolicy> {
    const response = await this.client.get<GovernancePolicy>(
      `${this.basePath}/policies/${policyId}`
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Governance policy not found: ${policyId}`);
    }

    return response.data;
  }

  // ===========================================================================
  // Escalation Paths
  // ===========================================================================

  /**
   * Lists escalation paths
   */
  async listEscalationPaths(): Promise<EscalationPath[]> {
    const response = await this.client.get<EscalationPath[]>(`${this.basePath}/escalation-paths`);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list escalation paths');
    }

    return response.data;
  }

  /**
   * Gets an escalation path by ID
   */
  async getEscalationPath(pathId: string): Promise<EscalationPath> {
    const response = await this.client.get<EscalationPath>(
      `${this.basePath}/escalation-paths/${pathId}`
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Escalation path not found: ${pathId}`);
    }

    return response.data;
  }

  // ===========================================================================
  // Constitutional Validation
  // ===========================================================================

  /**
   * Validates an action against constitutional principles
   */
  async validateConstitutional(params: {
    agentId: string;
    action: string;
    context: Record<string, unknown>;
    metadata?: Record<string, unknown>;
  }): Promise<{
    compliant: boolean;
    violations: Array<{
      principle: string;
      severity: 'warning' | 'violation' | 'critical';
      description: string;
    }>;
    recommendations: string[];
    requiresApproval: boolean;
    constitutionalHash: string;
  }> {
    const response = await this.client.post<{
      compliant: boolean;
      violations: Array<{
        principle: string;
        severity: 'warning' | 'violation' | 'critical';
        description: string;
      }>;
      recommendations: string[];
      requiresApproval: boolean;
      constitutionalHash: string;
    }>(`${this.basePath}/constitutional/validate`, {
      ...params,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to validate constitutional compliance');
    }

    return response.data;
  }

  /**
   * Gets constitutional principles
   */
  async getConstitutionalPrinciples(): Promise<Array<{
    id: string;
    name: string;
    description: string;
    category: string;
    priority: number;
    constitutionalHash: string;
  }>> {
    const response = await this.client.get<Array<{
      id: string;
      name: string;
      description: string;
      category: string;
      priority: number;
      constitutionalHash: string;
    }>>(`${this.basePath}/constitutional/principles`);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get constitutional principles');
    }

    return response.data;
  }

  // ===========================================================================
  // Metrics & Analytics
  // ===========================================================================

  /**
   * Gets governance metrics
   */
  async getMetrics(params?: {
    startDate?: string;
    endDate?: string;
    policyId?: string;
  }): Promise<GovernanceMetrics> {
    const response = await this.client.get<GovernanceMetrics>(
      `${this.basePath}/metrics`,
      params
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get governance metrics');
    }

    return response.data;
  }

  /**
   * Gets governance dashboard
   */
  async getDashboard(): Promise<{
    pendingApprovals: number;
    recentDecisions: GovernanceDecision[];
    complianceScore: number;
    activeAlerts: Array<{
      id: string;
      severity: 'low' | 'medium' | 'high' | 'critical';
      message: string;
      timestamp: string;
    }>;
    constitutionalHash: string;
  }> {
    const response = await this.client.get<{
      pendingApprovals: number;
      recentDecisions: GovernanceDecision[];
      complianceScore: number;
      activeAlerts: Array<{
        id: string;
        severity: 'low' | 'medium' | 'high' | 'critical';
        message: string;
        timestamp: string;
      }>;
      constitutionalHash: string;
    }>(`${this.basePath}/dashboard`);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get governance dashboard');
    }

    return response.data;
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  private validateApprovalRequest(data: unknown): ApprovalRequest {
    const result = ApprovalRequestSchema.safeParse(data);
    if (!result.success) {
      throw new ValidationError('Invalid approval request data', {
        validation: result.error.errors.map((e) => e.message),
      });
    }
    return result.data;
  }

  private validateDecision(data: unknown): GovernanceDecision {
    const result = GovernanceDecisionSchema.safeParse(data);
    if (!result.success) {
      throw new ValidationError('Invalid governance decision data', {
        validation: result.error.errors.map((e) => e.message),
      });
    }
    return result.data;
  }
}

File: src/core/sdk/python/acgs2_sdk/services/governance.py
"""
ACGS-2 Governance Service
Constitutional Hash: cdd01ef066bc6cf2
"""

from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any, Dict, List, Union

try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = Union[str, int, float, bool, None, Dict[str, Any], List[Any]]  # type: ignore[misc]
    JSONDict = Dict[str, JSONValue]  # type: ignore[misc]

from acgs2_sdk.constants import CONSTITUTIONAL_HASH, GOVERNANCE_ENDPOINT
from acgs2_sdk.models import (
    ApprovalRequest,
    ApprovalStatus,
    CreateApprovalRequest,
    GovernanceDecision,
    PaginatedResponse,
    SubmitApprovalDecision,
)

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class GovernanceService:
    """Service for governance and approval workflows."""

    def __init__(self, client: "ACGS2Client") -> None:
        self._client = client
        self._base_path = GOVERNANCE_ENDPOINT

    async def create_approval_request(
        self,
        request: CreateApprovalRequest,
    ) -> ApprovalRequest:
        """Create an approval request."""
        data = await self._client.post(
            f"{self._base_path}/approvals",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return ApprovalRequest.model_validate(data.get("data", data))

    async def get_approval_request(self, request_id: str) -> ApprovalRequest:
        """Get an approval request by ID."""
        data = await self._client.get(f"{self._base_path}/approvals/{request_id}")
        return ApprovalRequest.model_validate(data.get("data", data))

    async def list_approval_requests(
        self,
        page: int = 1,
        page_size: int = 50,
        status: ApprovalStatus | None = None,
        requester_id: str | None = None,
        pending_for: str | None = None,
    ) -> PaginatedResponse[ApprovalRequest]:
        """List approval requests."""
        params: JSONDict = {"page": page, "pageSize": page_size}
        if status:
            params["status"] = status.value
        if requester_id:
            params["requesterId"] = requester_id
        if pending_for:
            params["pendingFor"] = pending_for

        data = await self._client.get(f"{self._base_path}/approvals", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[ApprovalRequest](
            data=[ApprovalRequest.model_validate(r) for r in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def submit_decision(
        self,
        request_id: str,
        decision: SubmitApprovalDecision,
    ) -> ApprovalRequest:
        """Submit an approval decision."""
        data = await self._client.post(
            f"{self._base_path}/approvals/{request_id}/decisions",
            json={
                **decision.model_dump(by_alias=True),
                "timestamp": datetime.now(UTC).isoformat(),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return ApprovalRequest.model_validate(data.get("data", data))

    async def escalate(
        self,
        request_id: str,
        reason: str,
    ) -> ApprovalRequest:
        """Escalate an approval request."""
        data = await self._client.post(
            f"{self._base_path}/approvals/{request_id}/escalate",
            json={"reason": reason},
        )
        return ApprovalRequest.model_validate(data.get("data", data))

    async def cancel_approval_request(
        self,
        request_id: str,
        reason: str | None = None,
    ) -> None:
        """Cancel an approval request."""
        await self._client.post(
            f"{self._base_path}/approvals/{request_id}/cancel",
            json={"reason": reason},
        )

    async def get_decision(self, decision_id: str) -> GovernanceDecision:
        """Get a governance decision by ID."""
        data = await self._client.get(f"{self._base_path}/decisions/{decision_id}")
        return GovernanceDecision.model_validate(data.get("data", data))

    async def list_decisions(
        self,
        page: int = 1,
        page_size: int = 50,
        decision: str | None = None,
        request_id: str | None = None,
        reviewer_id: str | None = None,
    ) -> PaginatedResponse[GovernanceDecision]:
        """List governance decisions."""
        params: JSONDict = {"page": page, "pageSize": page_size}
        if decision:
            params["decision"] = decision
        if request_id:
            params["requestId"] = request_id
        if reviewer_id:
            params["reviewerId"] = reviewer_id

        data = await self._client.get(f"{self._base_path}/decisions", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[GovernanceDecision](
            data=[GovernanceDecision.model_validate(d) for d in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def verify_decision_anchor(self, decision_id: str) -> JSONDict:
        """Verify a decision's blockchain anchor."""
        data = await self._client.get(f"{self._base_path}/decisions/{decision_id}/verify")
        return data.get("data", data)

    async def validate_constitutional(
        self,
        agent_id: str,
        action: str,
        context: JSONDict,
        metadata: JSONDict | None = None,
    ) -> JSONDict:
        """Validate an action against constitutional principles."""
        data = await self._client.post(
            f"{self._base_path}/constitutional/validate",
            json={
                "agentId": agent_id,
                "action": action,
                "context": context,
                "metadata": metadata,
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return data.get("data", data)

    async def get_constitutional_principles(self) -> list[JSONDict]:
        """Get constitutional principles."""
        data = await self._client.get(f"{self._base_path}/constitutional/principles")
        return data.get("data", [])

    async def get_metrics(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
        policy_id: str | None = None,
    ) -> JSONDict:
        """Get governance metrics."""
        params: JSONDict = {}
        if start_date:
            params["startDate"] = start_date
        if end_date:
            params["endDate"] = end_date
        if policy_id:
            params["policyId"] = policy_id

        data = await self._client.get(f"{self._base_path}/metrics", params=params)
        return data.get("data", data)

    async def get_dashboard(self) -> JSONDict:
        """Get governance dashboard data."""
        data = await self._client.get(f"{self._base_path}/dashboard")
        return data.get("data", data)
